### ISSUE TITLE
Google layers do not work in a layer group

### ISSUE THREAD

**Issue by mstenta** [mapgears/ol3-google-maps#285]
Broke this out into a separate issue from #37 

> If I put the Google layers in the root map layers collection, they work. But if I put them inside a LayerGroup (eg: "Base layers"), they do not work. You can toggle between them, so nothing in the code seems to break... you just see a white map screen when you toggle to a Google layer.

**Comment by mstenta** [comment]
I've spent some time investigating this, and I think I have a pretty good sense of where changes will be needed in the code to support this.

Basically, the library currently assumes that Google layers will *not* be nested inside a layer group.  So it only works when they are not nested.

Here are links to specific sections of code that I think need to be updated:

https://github.com/mapgears/ol3-google-maps/blob/3a3c82f3e2c287d14a9303f854e7cfe8af0dfc43/src/olgm/herald/Source.js#L66

https://github.com/mapgears/ol3-google-maps/blob/3a3c82f3e2c287d14a9303f854e7cfe8af0dfc43/src/olgm/herald/Layers.js#L167

https://github.com/mapgears/ol3-google-maps/blob/3a3c82f3e2c287d14a9303f854e7cfe8af0dfc43/src/olgm/herald/Layers.js#L189

https://github.com/mapgears/ol3-google-maps/blob/3a3c82f3e2c287d14a9303f854e7cfe8af0dfc43/src/olgm/herald/Layers.js#L410

**Comment by mstenta** [comment]
I made some progress on this before the holidays. I will commit my work to a new branch and open a pull request soon. I have Google layers working in layer groups, but there is still a bit more to do before this is "complete" and ready for review/merge. More soon...

**Comment by mstenta** [comment]
Pushed my work-in-progress to a new fork branch: https://github.com/mapgears/ol3-google-maps/compare/master...mstenta:layergroup

I made two changes, and now Google Maps layers are working inside a layer group:

- In `watchLayer_()` and `unwatchLayer_()` I added an `else if` to check if the layer is a `LayerGroup`. If so, it will recurse into it and run the watch/unwatch function on each layer within it. See https://github.com/mapgears/ol3-google-maps/commit/07cfb0611d1c3a63c1b4c5726c112ab675e47923
- In `toggleGoogleMaps_()` I changed the way that the "find top-most Google layer" code works. Instead of looking at all layers in `this.ol3map.getLayers().getArray()`, it only looks through layers that were previously discovered to be Google layers, which are stored in `this.googleLayers_`. See https://github.com/mapgears/ol3-google-maps/commit/10c898805ec7ee24f0d0121613c5f3c7f7bf658f

These two changes cover the activation and deactivation of Google layers, which covers 3 of the 4 code sections I point to in my comment above (https://github.com/mapgears/ol3-google-maps/issues/285#issuecomment-568072168) - specifically the ones in `ol3-google-maps/src/olgm/herald/Layers.js`.

It does not yet cover the code section in `ol3-google-maps/src/olgm/herald/Styles.js`, in the `findIndex(layer)` method, which is used in the following two places:

1. `ImageWMSSourceHerald.orderLayers()`: https://github.com/mapgears/ol3-google-maps/blob/3a3c82f3e2c287d14a9303f854e7cfe8af0dfc43/src/olgm/herald/ImageWMSSource.js#L380
2. `TileSourceHerald.orderLayers()`: https://github.com/mapgears/ol3-google-maps/blob/3a3c82f3e2c287d14a9303f854e7cfe8af0dfc43/src/olgm/herald/TileSource.js#L339

So, that means that maps with Image WMS sources or Tile sources may not behave properly. This still needs to be resolved.

Apart from that, I also thought of another consideration: in `LayersHerald.activate()` there are listeners being added to watch for layers being added/removed from the map (https://github.com/mapgears/ol3-google-maps/blob/3a3c82f3e2c287d14a9303f854e7cfe8af0dfc43/src/olgm/herald/Layers.js#L175). This is not covering layers being added/removed from layer groups, however. So if layers are added/removed from the map AFTER Google Maps is activated, they will not be watched/unwatched.

So to summarize, there are two remaining pieces to solve:

1. The `findIndex(layer)` method, used by `ImageWMSSourceHerald` and `TileSourceHerald`.
2. Layers that are added to/removed from layer groups after Google Maps has been activated.

That said, if your map does not need either of those, then the first two commits I made might be enough.

**Comment by mstenta** [comment]
I ran into some issues testing this with vector layers. Seeing the following error in console when I switch between Google and non-Google base layers:

`TypeError: this.cache_[index] is undefined` coming from `unwatchFeature_()` in `herald/VectorFeature.js`.

I discovered that I can work around this by telling OLGoogleMaps NOT to watch vector layers, which also lets me use clusters (https://github.com/mapgears/ol3-google-maps/issues/52#issuecomment-570809053).

So for now I'm using the two commits for this issue, and disabling vector layer management in OLGoogleMaps. Seems to be working so far...

**Comment by adube** [comment]
Is there a way to reproduce the error you're describing using one of the current examples?

**Comment by mstenta** [comment]
I do have a commit in my fork that modifies the (newly restored) layer switcher example and puts the two layers into a layer group: https://github.com/mapgears/ol3-google-maps/commit/2c8442808d693f673be07f793041d839ed47bf74

If you add a vector layer to that I think you should see the error.

**Comment by adube** [comment]
Okay, let me know if I understood the issue correctly: if a vector layer is added in a group, then an error gets thrown while enabling/disabling OLGM.  Is that correct?

**Comment by mstenta** [comment]
There are actually three things I'm aware of that need to be addressed for this issue. See [this comment](https://github.com/mapgears/ol3-google-maps/issues/285#issuecomment-570804165) for the first 2. The third is the `VectorFeature` error.

> if a vector layer is added in a group, then an error gets thrown while enabling/disabling OLGM. Is that correct?

Yes, I believe that's correct.

**Comment by adube** [comment]
The OLGM library is now deprecated.  See: https://github.com/mapgears/ol3-google-maps/pull/352
